02-microarray/pathway-analysis_microarray_03_gsva.Rmd

```diff
@@ -0,0 +1,428 @@
+
+The function we will use to map from Ensembl gene IDs to Entrez gene IDs is called `mapIds()`.
+
+Let's create a data frame that shows the mapped Entrez IDs along with the gene expression values for the respective Ensembl IDs.
+
+```{r}
+# First let's create a mapped data frame we can join to the gene expression values
+mapped_df <- data.frame(
+  "entrez_id" = mapIds(
+    org.Hs.eg.db, # Replace with annotation package for the organism relevant to your data
+    keys = df$Gene,
+    column = "ENTREZID", # Replace with the type of gene identifiers you would like to map to
+    keytype = "ENSEMBL", # Replace with the type of gene identifiers in your data
+    multiVals = "first" # This will keep only the first mapped value for each Ensembl ID
+  )
+) %>%
+  # If an Ensembl gene identifier doesn't map to a Entrez gene identifier, drop that
+  # from the data frame
+  dplyr::filter(!is.na(entrez_id)) %>%
+  # Make an `Ensembl` column to store the rownames
+  tibble::rownames_to_column("Ensembl") %>%
+  # Now let's join the rest of the expression data
+  dplyr::inner_join(df, by = c("Ensembl" = "Gene"))
+```
+
+This `1:many mapping between keys and columns` message means that some Ensembl gene identifiers map to multiple Entrez IDs.
+In this case, it's also possible that a Entrez ID will map to multiple Ensembl IDs.
+For the purpose of performing GSEA later in this notebook, we keep only the first mapped IDs.
+For more about how to explore this, take a look at our [microarray gene ID conversion example](https://alexslemonade.github.io/refinebio-examples/02-microarray/gene-id-annotation_microarray_01_ensembl.html).
+
+Let's see a preview of `mapped_df`.
+
+```{r}
+head(mapped_df)
+```
+
+We will want to keep in mind that GSVA requires that data is in a matrix with the gene identifiers as rownames.
+In order to successfully turn our data frame into a matrix, we will need to ensure that we do not have any duplicate gene identifiers.
+
+Let's check to see if we have any Entrez IDs that mapped to multiple Ensembl IDs.
+
+```{r}
+any(duplicated(mapped_df$entrez_id))
+```
+
+Looks like we do have duplicated Entrez IDs.
+Let's find out which Entrez IDs have been duplicated.
+
+```{r}
+dup_entrez_ids <- mapped_df %>%
+  dplyr::filter(duplicated(entrez_id)) %>%
+  dplyr::pull(entrez_id)
+
+dup_entrez_ids
+```
+
+Now let's take a look at the rows associated with the duplicated Entrez IDs.
+
+```{r}
+mapped_df %>%
+  dplyr::filter(entrez_id %in% dup_entrez_ids)
+```
+
+We can see that the associated values vary for each row.
+
+As we mentioned earlier, we will want to remove duplicated gene identifiers in preparation for the GSVA steps later, so let's keep the Entrez IDs associated with the maximum expression value per sample.
```

Comment:
We may also want to remind the user that these values are representing probe IDs ultimately, even though the Ensembl column shows transcript IDs.  This is a tad different philosophically from RNA-seq where we've read the transcript itself; here we really are just dealing with probe differences here.

Reply:
Yes, but it's a little bit more nuanced that that because we're using BrainArray with SCAN. Quoting the [`SCAN.UPC`](https://bioconductor.org/packages/release/bioc/vignettes/SCAN.UPC/inst/doc/SCAN.vignette.pdf) documentation:

> By default,SCANmaps Affymetrix probes to “probeset” identifiers provided by themanufacturer.   However,  these  mappings  may  be  outdated  and  include  problematicprobes (for example, those that may cross hybridize).  In addition, multiple probesetsmay be assigned to a single gene.  As an alternative, the BrainArray resource (seehttp://brainarray.mbni.med.umich.edu/Brainarray/Database/CustomCDF/genomic_curated_CDF.asp) provides regularly updated mappings that exclude problematic probes and mapto genes rather than probesets.  

The target audience for this probably doesn't need all that information, so you can use terms like Ensembl gene identifier (not transcript identifiers, those would start with `ENST`).

Reply:
My use of the word transcript is confusing/vague/a bit wrong. I meant to differentiate it from rna-seq reads, but no, these are not transcript IDs.